php: Add runnable tests

RemcoSmitsDev opened the pull request:
### This pull request adds the following:
- Missing mapping for the `yield` keyword.
- Outline scheme for `describe`, `it` and `test` function_call_expressions (to support Pest runnable)
- Pest runnable support
- PHPUnit runnable support
- Task for running selected PHP code.


## Queries explanations

#### Query 1 (PHPUnit: Run specific method test):
1. Class is not abstract (because you cannot run tests from an abstract class)
2. Class has `Test` suffix
3. Method has public modifier(or no modifiers, default is public)
4. Method has `test` prefix

#### Query 2 (PHPUnit: Run specific method test with `@test` annotation):
1. Class is not abstract (because you cannot run tests from an abstract class)
2. Class has `Test` suffix
3. Method has public modifier(or no modifiers, default is public)
4. Method has `@test` annotation

#### Query 3 (PHPUnit: Run specific method test with `#[Test]` attribute):
1. Class is not abstract (because you cannot run tests from an abstract class)
2. Class has `Test` suffix
3. Method has public modifier(or no modifiers, default is public)
4. Method has `#[Test]` attribute

#### Query 4 (PHPUnit: Run all tests inside the class):
1. Class is not abstract (because you cannot run tests from an abstract class)
2. Class has `Test` suffix

#### Query 5 (Pest: Run function test)
1. Function expression has one of the following names: `describe`, `it` or `test`
2. Function expression first argument is a string

### **PHPUnit: Example for valid test class**
<img width="549" alt="Screenshot 2024-05-08 at 10 41 34" src="https://github.com/zed-industries/zed/assets/62463826/e84269de-4f53-410b-b93b-713f9448dc79">


### **PHPUnit: Example for invalid test class**
All the methods should be ignored because you cannot run tests on an abstract class.
<img width="608" alt="Screenshot 2024-05-07 at 22 28 57" src="https://github.com/zed-industries/zed/assets/62463826/8c6b3921-5266-4d88-ada5-5cd827bcf242">

### **Pest: Example**
https://github.com/zed-industries/zed/assets/62463826/bce133eb-0a6f-4ca2-9739-12d9169bb9d6

You should now see all your **Pest** tests inside the buffer symbols modal.
![Screenshot 2024-05-08 at 22 51 25](https://github.com/zed-industries/zed/assets/62463826/9c818b74-383c-45e5-9b41-8dec92759a14)

Release Notes:

- Added test runnable detection for PHP (PHPUnit & Pest).
- Added task for running selected PHP code.
- Added `describe`, `test` and `it` functions to buffer symbols, to support Pest runnable.
- Added `yield` keyword to PHP keyword mapping.

RemcoSmitsDev committed:
Add missing keyword mapping for `yield`

RemcoSmitsDev committed:
Add runnables query + task for running phpunit tests

claytonrcarter commented on extensions/php/languages/php/runnables.scm:
```diff
@@ -0,0 +1,31 @@
+(
+    ; Class that follow the naming convention of PHPUnit test classes
+    ; and that doesn't have the abstract modifier
+    ; and have a method that follow the naming convention of PHPUnit test methods
+    ; and the method is public
+    (class_declaration
+        modifier: (_)? @modifier
+        (#not-eq? @modifier "abstract")
+        name: (_) @name
+        (#match? @name ".*Test")
+        body: (declaration_list
+            (method_declaration
+                (visibility_modifier)? @visibility
+                (#eq? @visibility "public")
+                name: (_) @run
+                (#match? @run "test.*")
+            )
+        )
+    )
+) @php-test
```
It would be great to also support tests that are annotated with `@test` phpdoc comments: https://docs.phpunit.de/en/11.1/annotations.html#test

This can be added with a query like this (note `(#match? @_comment ".* @test .*")`)
```lisp
; Methods withs @test annotations
(class_declaration
    name: (_) @_class
	body: (declaration_list
        ((comment) @_comment
         .
         (method_declaration
            name: (_) @run)))

    (#match? @_class ".*Test$")
    (#match? @_comment ".* @test .*")
) @php-test
```

RemcoSmitsDev replied:
I thought they dropped support for docblock annotations. I will update the PR to support them!

claytonrcarter commented on extensions/php/languages/php/tasks.json:
```diff
@@ -0,0 +1,8 @@
+[
+  {
+    "label": "php test $ZED_SYMBOL",
+    "command": "vendor/bin/phpunit",
+    "args": ["--filter=$ZED_SYMBOL"],
+    "tags": ["php-test"]
```
*nitpick*: the [recently added](4eca7875ae25469c5a43f3f73f7160c861cba3c8) gleam test runner uses the label `gleam test ...` but, in that case, the command that is actually being run is `gleam test ...`. In this case, the label may be more accurately named `phpunit $ZED_SYMBOL`

Ditto for the tag name: perhaps `phpunit-test` would be a better tag name, in case someone wants to later add support for [alternate PHP test runners](https://pestphp.com/).

RemcoSmitsDev replied:
Agree, will change that!

claytonrcarter reviewed (commented):
Cool! Just leaving a few comments based on a version that I was working on before I saw that someone had beat me to it! 😄

osiewicz reviewed (approved):
LGTM, thanks Remco :)

RemcoSmitsDev commented:
> LGTM, thanks Remco :)

I gonna add pest & docblock annotation support aswell. Working on it right now!

RemcoSmitsDev committed:
Rename php to phpunit

RemcoSmitsDev committed:
Add support for @test annotation

RemcoSmitsDev commented on extensions/php/languages/php/runnables.scm:
```diff
@@ -0,0 +1,54 @@
+            (method_declaration
+                (visibility_modifier)? @visibility
+                (#eq? @visibility "public")
+                name: (_) @run
```
@osiewicz Is there away to match the method name and ignore if its start with test? Because we now get the following error. 

```log
[2024-05-08T10:07:18+02:00 ERROR editor] multiple different run targets found on a single line, only the last target will be rendered
```

This happends if your method looks like this:

```php
    /**
     * @test
     */
    public function testValidateCLass()
    {
        $this->assertTrue(true);
    }
```

osiewicz replied:
Can't we just modify the regex on line 31 to be `(#match? @name ".+Test")`?

osiewicz replied:
How about
```suggestion
                name: (_) @run
                (#not-match? @run "test.*")
```
?

RemcoSmitsDev replied:
That works😅

osiewicz replied:
Tbh though, it looks like something we should fix on Zed side and not here. I can't really squeeze it into todays Preview, but it would be a bogus limitation if we could only render tests from a single framework/line.

I'd say: let's proceed with that `not-match` approach and later on get rid of it once we can support multiple tests on the same line.

RemcoSmitsDev replied:
Yeah, that's what I thought. It looks good to me now, will make a follow-up to support pest there are a little bit more queries to support all the pest conventions.

RemcoSmitsDev committed:
Make regex more strict

RemcoSmitsDev committed:
Fix render error for multiple targets

Error: multiple different run targets found on a single line, only the last target will be rendered

claytonrcarter commented on extensions/php/languages/php/runnables.scm:
```diff
@@ -0,0 +1,55 @@
+        (#match? @name ".*Test$")
+        body: (declaration_list
+            ((comment) @comment
+                (#match? @comment ".*@test.*")
```
Minor change: this should probably only match `@test` so that things like `@testdox`, `@testWith` or (eg) `@test-disabled` do not mark the annotated method as a test. (I checked and `@testdox` and `@testWith` do not imply that it's a test on their own; they must be paired with `@test` or `test_...`.)
 I suspect that something like this would take care of it for most use cases:
```suggestion
                (#match? @comment ".*@test\s.*")
```

Fun fact, I played with phpunit a little and found the following:
```php
/** @test */    ← works
/** @test*/     ← works
/**@test */     ← does not work
/** aaa@test */ ← works
/** @test       ← works (trimmed line/newline terminated)
 */
```
:shrug:

claytonrcarter commented:
> Ahhh sorry man, didn't know that. I hope this satisfies your needs!

It definitely does. No worries here! Thank you!

RemcoSmitsDev committed:
Fix don't match any annotation that starts with @test

claytonrcarter reviewed (approved):
Looks great! Thank you to doing this. It's certainly more robust that what I had hacked together! 😆

RemcoSmitsDev commented:
@osiewicz Can you help me out, I have the following 2 queries. They seem to work because I can see the runnable icon before the test functions. But when I click on them nothing happens, what am I missing here? Or not understanding correctly? It seems like it cannot get the value correctly from the query.

<img width="431" alt="Screenshot 2024-05-08 at 16 06 07" src="https://github.com/zed-industries/zed/assets/62463826/28ab2eb3-d0a0-483b-9e12-44128f8a51a7">

This query support non-chaining methods:
```php
test("example", function () {
    expect(true)->toBeTrue();
});
```
```scm
(
    (expression_statement
        (function_call_expression
            function: (_) @name
            (#any-of? @name "it" "test" "describe")
            arguments: (arguments
                (argument
                    (encapsed_string (string_value)) @run
                )
            )
        )
    )
) @pest-test
```

This query support chaining methods:
```php
test("example", function () {
    expect(true)->toBeTrue();
})->with(["1", "2"]);
```
```scm
(
    (expression_statement
        (member_call_expression
            object: (function_call_expression
                function: (_) @name
                (#any-of? @name "it" "test" "describe")
                arguments: (arguments
                    (argument
                        (encapsed_string (string_value)) @run
                    )
                )
            )
        )
    )
) @pest-test
```

RemcoSmitsDev committed:
Add `describe`, `it` and `test` function call expression to PHP outline scheme

This will allow to run the Pest runnable. Because before this change
These function call expressions were not recognized as symbols.

This resulted in that the Symbol key inside the task context not being set.

RemcoSmitsDev committed:
Fix script was not run in correct location

RemcoSmitsDev committed:
Fix only run test with the symbol name inside the current file

RemcoSmitsDev committed:
Add support for pest runnable

RemcoSmitsDev commented:
While digging inside the runnable/task code, looking why clicking the runnable icon inside the gutter for **Pest** runnable didn't open a new terminal panel to run the tests in. I finally found the issue that not only applies to **Pest** runnable.

The issue is that the following code does not contain any **symbols** according to the Zed PHP outline scheme.
```php
<?php

test("example", function () {
    expect(true)->toBeTrue();
});

describe("tests", function () {
    test("test1", function () {
        expect(true)->toBeTrue();
    });

    test("test2", function () {
        expect(true)->toBeTrue();
    });
});

``` 

So when the task context is getting build inside `task_context::build_context` (line: 53). We ask for all the **symbols**, but this will give us an empty **Vec** back. Because we have an empty **Vec** we cannot determine what the **symbol** for the task context should be.

So if we go higher up the stack we are trying to build a task from the task template, which could not be done because the **symbol** information does not exist inside the context which is required for the **Pest** task template.

---

So I was talking to @osiewicz to explain what I found, why the **Pest** runnable did not work. So I came up with a fix for this issue by adding the following code to the **PHP** outline scheme:

```scm
; Add support for Pest runnable (non chainable method)
(expression_statement
    (function_call_expression
        function: (_) @context
        (#any-of? @context "it" "test" "describe")
        arguments: (arguments
            (argument
                (encapsed_string (string_value) @name)
            )
        )
    )
) @item

; Add support for Pest runnable (chainable method)
(expression_statement
    (member_call_expression
        object: (function_call_expression
            function: (_) @context
            (#any-of? @context "it" "test" "describe")
            arguments: (arguments
                (argument
                    (encapsed_string (string_value) @name)
                )
            )
        )
    )
) @item
```

After adding this to the outline scheme, I got the **Pest** runnable too work😀 That being said, as I already mentioned we are going to have this issue with more languages. So @osiewicz came up with an idea to store the query captures inside the **Env** key of the task context. So we could use them instead of having to add these expressions to the outline scheme.

---

### Note
We have to think about how we want to solve this in the feature. For now, having them inside the outline scheme is fine.

RemcoSmitsDev committed:
Allow multiple chainable methods for Pest runnable

This allows people to use the following code:

```php
it("has a name")
    ->expect(fn() => User::create(["name" => "Nuno Maduro"])->name)
    ->toBe("Nuno Maduro");

--- And ---

describe("tests in describe", function () {
    test("test1", function () {
        expect(true)->toBeTrue();
    });

    test("test2", function () {
        expect(true)->toBeTrue();
    });
})->with(["1", "2"])->group("group1");
```

RemcoSmitsDev committed:
Modify outline scheme for Pest runnable chainable methods

This allows multiple chainable methods to be run in a single test

RemcoSmitsDev committed:
Clean up runnable queries

RemcoSmitsDev committed:
Move Pest runnable query/scheme to single statement

We have to do this because we cannot specify dynamic child node matching for more than one node.

If you would have more than one chained method the previous version did not work.

NOTE: This will also match test("asdf") inside function/method declarations..

RemcoSmitsDev commented:
@osiewicz one thing I'm not happy about yet is that the Pest query/outline scheme is not specific enough. This results in invalid results that should not happen, but could.

<img width="750" alt="Screenshot 2024-05-09 at 11 45 33" src="https://github.com/zed-industries/zed/assets/62463826/c1d39895-90b8-4207-9895-1500cda7a9a9">

The original query/scheme that I made prevented this, but did not support more than one chained method.
This is because the more chained methods you have, the more **member_call_expressions** you have as child nodes. I could not fixture out how I could write a query/scheme that has dynamic wildcard for more than one child node using the `(_)` syntax.

So ideally I want to wildcard all the **member_call_expressions** that are inside the **expression_statement**. Is there away to do this?

```scm
```

<img width="373" alt="Screenshot 2024-05-09 at 11 43 43" src="https://github.com/zed-industries/zed/assets/62463826/dcee35f7-62ff-4275-b612-5bad30fd7d9a">

osiewicz commented:
Hey, wanna pair on it sometime later in the day? E.g. at 5 PM (we're in the same timezone).

RemcoSmitsDev commented:
Yeah sure, sounds good!

claytonrcarter commented on extensions/php/languages/php/runnables.scm:
```diff
@@ -0,0 +1,61 @@
+    )
+) @phpunit-test
+
+; Class that follow the naming convention of PHPUnit test classes
+; and that doesn't have the abstract modifier
+; and have a method that has the @test annotation
+; and the method is public
+(class_declaration
+    modifier: (_)? @modifier
+    (#not-eq? @modifier "abstract")
+    name: (_) @name
+    (#match? @name ".*Test$")
+    body: (declaration_list
+        ((comment) @comment
+            (#match? @comment ".*@test\\b.*")
+        .
+        (method_declaration
+            (visibility_modifier)? @visibility
+            (#eq? @visibility "public")
+            name: (_) @run
+            (#not-match? @run "^test.*")
+        ))
+    )
+) @phpunit-test
```
I hope my mentioning Pest didn't make this more complicated than it needed to be! I'm sorry if it did. I've not used pest, but as far as the phpunit support, this LGTM!

> agree that \b it better is in this case

Yes, I think that `\b` is the way to go! Good call 👍 

> deprecate docblocks in PHP and just use attributes

So ... I wasn't aware of this ... and it's definitely something that can be left for a followup PR to not further complicate this ... *but* ... your comment sent me looking and apparently phpunit does have YAWTMAT (yet another way to mark a test), using attributes: https://docs.phpunit.de/en/11.1/attributes.html#test I've not used it (wasn't even aware of it!) but I think that adding these lines to the `method_declaration` in the other query (for `test*` methods) may work:
```lisp
attributes: (attribute_list (attribute_group (attribute (name) @attr)))?
(#eq? @attr "Test")
```

RemcoSmitsDev replied:
I heard they were planning to add attribute support, but didn't know it was already out. But thanks for pointing it out! Will add support for it.

RemcoSmitsDev replied:
No, it's fine!, I was already planning to make a follow-up for **Pest** support. But when you mentioned it I thought let just add it inside this PR. Yeah, it is more complicated than I thought, tho.

claytonrcarter commented:
> This results in invalid results that should not happen, but could.

Can Pest tests be defined in classes, or are they only correct when defined at the "top level" of the file/script? If so, could it work to query for any `function_call_expression` that are *not* in a `class_declaration`?

RemcoSmitsDev commented:
**Pest** does not support the following methods `it`, `test` and `describe` inside classes or function declarations. Because these functions are expression that way, they register as a test case. You if you would define them in any declaration kind it's not getting registered as a test case.

Yeah, we have to do something like that. But I'm not sure how to do that.

RemcoSmitsDev committed:
Add support for PHPUnit #[Test] attributes

claytonrcarter commented:
> Pest does not support the following methods it, test and describe inside classes or function declarations

If this is the case, would it work to wrap the query in `(program (expression_statement ...))` to only match expressions at the "top level" and not, like in your example, within a class method? Hmm, though I guess that wouldn't work if the tests were wrapped in `if (...) { }` or `foreach (...) { }`?

Is there a tree-sitter way to query for something that does not have a particular node as an ancestor?

RemcoSmitsDev committed:
Add _ to all not needed captures

RemcoSmitsDev commented:
Yeah, adding the program as top level node inside the query fixes the issue that it matches also expressions inside the function/class declarations. The query looks like this:

```scm
; Match non chained methods for Pest test
(program
    (expression_statement
        (function_call_expression
            function: (_) @_name
            (#any-of? @_name "it" "test" "describe")
            arguments: (arguments
                (argument
                    (encapsed_string (string_value) @run)
                )
            )
        )
    )
) @pest-test
```

RemcoSmitsDev committed:
Fix filter for names with spaces

RemcoSmitsDev committed:
Pest: Only match first argument

mansoorkhan96 commented:
@RemcoSmitsDev possible to add support for Laravel Dusk?

RemcoSmitsDev commented:
@mansoorkhan96 You can make that possible by your self, when this pull request is merged by adding the following task to your `~/.config/zed/tasks.json`

```json
{
  "label": "laravel dusk test $ZED_SYMBOL",
  "command": "php artisan dusk",
  "args": ["--filter $ZED_SYMBOL $ZED_FILE"],
  "tags": ["phpunit-test", "pest-test"]
}
```

osiewicz commented:
@RemcoSmitsDev  I've merged main into this branch btw

RemcoSmitsDev committed:
Add task for executing selection

osiewicz commented:
Thanks!
